### PR TITLE
[FLINK-32218][Connector/Kinesis] Add support for parent-child shard ordering to Kinesis streams source

### PR DESCRIPTION
**Pull request by z3d1k**
## Purpose of the change

Adding support for parent-child shard ordering when reading from Kinesis streams.
When enabled, source enumerator will assign split that correspond to a shard only when parent shard either finished, expired, or don't exists.

## Verifying this change

Added unit tests.

## Significant changes
*(Please check any boxes [x] if the answer is "yes". You can first publish the PR and check them afterwards, for convenience.)*
- [ ] Dependencies have been added or upgraded
- [ ] Public API has been changed (Public API is any class annotated with `@Public(Evolving)`)
- [x] Serializers have been changed
- [x] New feature has been introduced
  - If yes, how is this documented? JavaDocs, connector documentation will be written in separate task [(FLINK-31989)](https://issues.apache.org/jira/browse/FLINK-31989)

### CODE REVIEW

**Comment by hlteoh37** [comment]
> New feature has been introduced
    If yes, how is this documented? JavaDocs

We will also need to document this on Flink docs - let's link the documentation update JIRA

**Comment by z3d1k** [comment]
> We will also need to document this on Flink docs - let's link the documentation update JIRA

Linked documentation Jira.

File: flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/KinesisStreamsSourceEnumeratorStateSerializer.java [hunk 1]
```diff
@@ -109,13 +110,22 @@ public KinesisStreamsSourceEnumeratorState deserialize(
...
             for (int i = 0; i < numUnassignedSplits; i++) {
                 int serializedLength = in.readInt();
                 byte[] serializedSplit = new byte[serializedLength];
                 if (in.read(serializedSplit) != -1) {
+                    KinesisShardSplit deserializedSplit =
+                            splitSerializer.deserialize(splitSerializerVersion, serializedSplit);
+                    SplitAssignmentStatus assignmentStatus = SplitAssignmentStatus.UNASSIGNED;
+                    if (version == CURRENT_VERSION) {
+                        assignmentStatus = SplitAssignmentStatus.fromStatusCode(in.readInt());
+                    }
```
Comment (hlteoh37) [inline flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/KinesisStreamsSourceEnumeratorStateSerializer.java:124]: for previous serializer version, we should be safe to assume unassigned split right?

File: flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/SplitAssignmentStatus.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+    public static SplitAssignmentStatus fromStatusCode(int statusCode) {
+        for (SplitAssignmentStatus status : SplitAssignmentStatus.values()) {
+            if (status.getStatusCode() == statusCode) {
+                return status;
+            }
+        }
+
+        throw new IllegalArgumentException("Unknown status code: " + statusCode);
+    }
```
Comment (hlteoh37) [inline flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/SplitAssignmentStatus.java:31]: Noted that we have set up the serialization arrangement for enums correctly so that we can add more enums safely in the future :)

File: flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/tracker/SplitTracker.java [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    /**
+     * Flag controlling if tracker should wait before all parent splits will be completed before
+     * assigning split to readers.
+     */
+    private final boolean preserveShardOrdering;
+
+    /** Map of all discovered splits that have not been completed. */
+    private final Map<String, KinesisShardSplit> knownSplits = new ConcurrentHashMap<>();
+
+    /** Set of currently assigned split id. */
```
Comment (hlteoh37) [inline flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/tracker/SplitTracker.java:29]: nit: currently assigned splits

File: flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/tracker/SplitTracker.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+@Internal
+public class SplitTracker {
+    /**
+     * Flag controlling if tracker should wait before all parent splits will be completed before
+     * assigning split to readers.
+     */
+    private final boolean preserveShardOrdering;
+
+    /** Map of all discovered splits that have not been completed. */
+    private final Map<String, KinesisShardSplit> knownSplits = new ConcurrentHashMap<>();
```
Comment (hlteoh37) [inline flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/tracker/SplitTracker.java:27]: Can we add details of what the key is in the doc?

File: flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/tracker/SplitTracker.java [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    }
+
+    private boolean verifyAllParentSplitsAreFinished(KinesisShardSplit split) {
+        boolean allParentsFinished = true;
+        for (String parentSplitId : split.getParentShardIds()) {
+            allParentsFinished = allParentsFinished && isFinished(parentSplitId);
+        }
+
+        return allParentsFinished;
+    }
```
Comment (hlteoh37) [inline flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/tracker/SplitTracker.java:142]: Might be good to leave a comment/tests to validate the assumptions we have made here, that this logic only works because parent splits are always by definition existing in the set of known splits.

File: flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/tracker/SplitTracker.java [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+                                    isAssigned(split.splitId())
+                                            ? SplitAssignmentStatus.ASSIGNED
+                                            : SplitAssignmentStatus.UNASSIGNED;
+                            return new KinesisShardSplitWithAssignmentStatus(
+                                    split, assignmentStatus);
+                        })
+                .collect(Collectors.toList());
+    }
+
+    private boolean verifyAllParentSplitsAreFinished(KinesisShardSplit split) {
```
Comment (hlteoh37) [inline flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/tracker/SplitTracker.java:135]: Noted that this implementation also covers the edge case, when we restore from a stale snapshot that is taken when not consuming at the tip of stream. Let's make sure we have a test case for this!

### Edge case 1
The edge case I'm thinking of:
```
A ---\   /--- E
       C 
B ---/   \--- F
```
Above I show a shard that was merged, then split. Although this is unrealistic in a 2 shard system, it represents a situation that can easily occur in large shard systems.

First we imagine a snapshot is taken when stream is aware of A,B,C, but currently consuming from A+B.
```
Current stream condition:
A ---\   
      C 
B ---/   

Enumerator state:
knownSplits: A,B,C
assignedSplits: A,B
```

Ok imagine now we keep this snapshot until the stream has expired shards A + B + C

```
Current stream condition:
  /--- E
C (expired)
  \--- F
```

Now, what will happen?
1. We discover splits E + F.
```
Enumerator state:
knownSplits: C,E (waiting on C), F (waiting on C)
assignedSplits: A,B
```
2. Assigned splits A + B will be marked as finished because they are expired. Assumed behavior of the SplitReader.

```
Enumerator state:
knownSplits: C,E (waiting on C), F (waiting on C)
assignedSplits: 
```
3. Enumerator will assign split C.
4. Split C will be marked as finished because they are expired. (Once again assumed behavior of SplitReader)
5. We will then start consuming from E + F.



### Edge case 2

```
A ---\   /--- E ---\
       C            G
B ---/   \--- F ---/
```


First we imagine a snapshot is taken when stream is aware of A,B,C, but currently consuming from A+B.
```
Current stream condition:
A ---\   
      C 
B ---/   

Enumerator state:
knownSplits: A,B,C
assignedSplits: A,B
```

Then we keep snapshot until it is stale, and shards A,B,C,E,F have expired.

```
Current stream condition:
E (expired) ---\
                G
F (expired) ---/

Enumerator state:
knownSplits: A,B,C
assignedSplits: A,B
```

Now, what will happen?
1. We discover splits G.
```
Enumerator state:
knownSplits: A,B,C,G(waiting on E,F)
assignedSplits: A,B
```
2. Assigned splits A + B will be marked as finished because they are expired. Assumed behavior of the SplitReader.

```
Enumerator state:
knownSplits: C,G(waiting on E,F)
assignedSplits: 
```
3. Enumerator will assign split C + G
4. Split C will be marked as finished because they are expired. (Once again assumed behavior of SplitReader). We will start consuming from G directly.

**Review by hlteoh37 — CHANGES_REQUESTED** [review]
Wow - nice and sleek implementation @z3d1k . Added some thoughts

File: flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/event/SplitsFinishedEvent.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+package org.apache.flink.connector.kinesis.source.event;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.connector.source.SourceEvent;
+
+import java.util.Set;
+
+/** Source event used by source reader to communicate that splits are finished to enumerator. */
+@Internal
+public class SplitsFinishedEvent implements SourceEvent {
```
Comment (hlteoh37) [inline flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/event/SplitsFinishedEvent.java:28]: Q: Do we meed to implement `hashcode` and `equals` here?

File: flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/proxy/ListShardsStartingPosition.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+import org.apache.flink.util.Preconditions;
+
+import software.amazon.awssdk.services.kinesis.model.ShardFilter;
+import software.amazon.awssdk.services.kinesis.model.ShardFilterType;
+
+import java.time.Instant;
+
+/** Starting position to perform list shard request. */
+@Internal
+public class ListShardsStartingPosition {
```
Comment (hlteoh37) [inline flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/proxy/ListShardsStartingPosition.java:31]: Nice! 😄

File: flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/reader/KinesisStreamsSourceReader.java [hunk 9]
```diff
@@ -64,6 +66,8 @@ public KinesisStreamsSourceReader(
     @Override
     protected void onSplitFinished(Map<String, KinesisShardSplitState> finishedSplitIds) {
         finishedSplitIds.keySet().forEach(this::unregisterShardMetricGroup);
+        context.sendSourceEventToCoordinator(
+                new SplitsFinishedEvent(new HashSet<>(finishedSplitIds.keySet())));
     }
```
Comment (hlteoh37) [inline flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/reader/KinesisStreamsSourceReader.java:71]: nit: can we swap the order? It would be better to send `SplitsFinishedEvent`, but miss unregister metric group, compared to the alternative.

File: flink-connector-aws/flink-connector-aws-kinesis-streams/src/test/java/org/apache/flink/connector/kinesis/source/enumerator/KinesisStreamsSourceEnumeratorStateSerializerTest.java [hunk 10]
```diff
@@ -145,6 +129,27 @@ void testSerializeWithTrailingBytes() throws Exception {
                 .withMessageContaining("Unexpected trailing bytes when deserializing.");
     }
 
+    private List<KinesisShardSplitWithAssignmentStatus> getSplits(
+            IntStream assignedShardIdRange, IntStream unassignedShardIdRange) {
```
Comment (hlteoh37) [inline flink-connector-aws/flink-connector-aws-kinesis-streams/src/test/java/org/apache/flink/connector/kinesis/source/enumerator/KinesisStreamsSourceEnumeratorStateSerializerTest.java:133]: Nice 😄

File: flink-connector-aws/flink-connector-aws-kinesis-streams/src/test/java/org/apache/flink/connector/kinesis/source/enumerator/tracker/SplitTrackerTest.java [hunk 11]
```diff
@@ -0,0 +1,371 @@
...
+        splitTracker.addSplits(Collections.singletonList(split));
+
+        List<KinesisShardSplit> pendingSplits = splitTracker.splitsAvailableForAssignment();
+
+        assertThat(pendingSplits).containsExactly(split);
+    }
+
+    @ParameterizedTest
+    @ValueSource(booleans = {true, false})
+    public void testStateSnapshot(boolean preserveShardOrdering) {
```
Comment (hlteoh37) [inline flink-connector-aws/flink-connector-aws-kinesis-streams/src/test/java/org/apache/flink/connector/kinesis/source/enumerator/tracker/SplitTrackerTest.java:59]: nit: Might help readability to add some comments around the test structure

`// Given mixture of assigned and unassigned splits`

`// When registered with split tracker`

`// Then snapshot state accurately reflects assigned and unassigned splits`

File: flink-connector-aws/flink-connector-aws-kinesis-streams/src/test/java/org/apache/flink/connector/kinesis/source/enumerator/tracker/SplitTrackerTest.java [hunk 12]
```diff
@@ -0,0 +1,371 @@
...
+
+        // Verify that state does not contain finished split
+        assertThat(splitTracker.snapshotState(0))
+                .doesNotContain(expectedSplitState.get(0))
+                .containsExactlyInAnyOrder(expectedSplitState.get(1), expectedSplitState.get(2));
+    }
+
+    @ParameterizedTest
+    @ValueSource(booleans = {true, false})
+    public void testSplitSnapshotMarkingSplitsAsAssignedCorrectlyMappedToState(
```
Comment (hlteoh37) [inline flink-connector-aws/flink-connector-aws-kinesis-streams/src/test/java/org/apache/flink/connector/kinesis/source/enumerator/tracker/SplitTrackerTest.java:140]: nit: This is already covered in `testStateSnapshot`

File: flink-connector-aws/flink-connector-aws-kinesis-streams/src/test/java/org/apache/flink/connector/kinesis/source/proxy/KinesisStreamProxyTest.java [hunk 13]
```diff
@@ -413,10 +414,15 @@ private List<Shard> getTestShards(final int startShardId, final int endShardId)
...
+            ShardFilter startingPosition = null;
+            if (startShardId != null) {
+                startingPosition =
+                        ListShardsStartingPosition.fromShardId(startShardId).getShardFilter();
+            }
             ListShardsRequest expectedReq =
                     ListShardsRequest.builder()
                             .streamARN(streamArn)
-                            .exclusiveStartShardId(startShardId)
+                            .shardFilter(startingPosition)
```
Comment (hlteoh37) [inline flink-connector-aws/flink-connector-aws-kinesis-streams/src/test/java/org/apache/flink/connector/kinesis/source/proxy/KinesisStreamProxyTest.java:425]: Do we need to add a test for startingPosition of AT_TIMESTAMP?

File: flink-connector-aws/flink-connector-aws-kinesis-streams/src/test/java/org/apache/flink/connector/kinesis/source/enumerator/KinesisStreamsSourceEnumeratorTest.java [hunk 14]
```diff
@@ -275,53 +262,8 @@ void testAddSplitsBackWithoutSplitIsNoOp() throws Throwable {
             // Given enumerator has no assigned splits
             // When we add splits back
             // Then handled gracefully with no exception thrown
```
Comment (hlteoh37) [inline flink-connector-aws/flink-connector-aws-kinesis-streams/src/test/java/org/apache/flink/connector/kinesis/source/enumerator/KinesisStreamsSourceEnumeratorTest.java:264]: This comment is no longer true

**Review by hlteoh37 — COMMENTED** [review]
Added some nits, Thanks @z3d1k

File: flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/KinesisStreamsSourceEnumerator.java [hunk 15]
```diff
@@ -143,17 +156,101 @@ public void addReader(int subtaskId) {
...
 
     @Override
     public void close() throws IOException {
         streamProxy.close();
     }
 
-    private List<KinesisShardSplit> initialDiscoverSplits() {
-        List<Shard> shards = streamProxy.listShards(streamArn, lastSeenShardId);
-        return mapToSplits(shards, sourceConfig.get(STREAM_INITIAL_POSITION));
+    @VisibleForTesting
```
Comment (karubian) [inline flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/KinesisStreamsSourceEnumerator.java:169]: Can we avoid these "VisibleForTesting" methods by modifying the TestKinesisStreamProxy?

File: flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/KinesisStreamsSourceEnumerator.java [hunk 16]
```diff
@@ -211,32 +301,33 @@ private void assignSplits(List<KinesisShardSplit> discoveredSplits, Throwable th
...
                     context.currentParallelism(),
                     context.registeredReaders().size());
-            unassignedSplits.addAll(discoveredSplits);
             return;
         }
 
+        assignSplits();
+    }
+
+    private void assignSplits() {
```
Comment (karubian) [inline flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/KinesisStreamsSourceEnumerator.java:318]: Can we rename this to a name with more context to emphasize the difference between enumerator and source reader?

File: flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/KinesisStreamsSourceEnumerator.java (hunk 15 above)
Comment (z3d1k) [reply flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/KinesisStreamsSourceEnumerator.java:169]: In this case logic does not fit in the StreamsProxy, but also needs to be tested.
Discussed with @hlteoh37 offline and we agreed that VisibleForTesting is ok here.

File: flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/KinesisStreamsSourceEnumerator.java (hunk 16 above)
Comment (z3d1k) [reply flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/KinesisStreamsSourceEnumerator.java:319]: This method is performing split assignment to readers, IMO name represents the action well.

Renamed method above to clearer represent meaning.

File: flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/KinesisShardSplitWithAssignmentStatus.java [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+    public KinesisShardSplit split() {
+        return kinesisShardSplit;
+    }
+
+    public SplitAssignmentStatus assignmentStatus() {
+        return splitAssignmentStatus;
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
[the remark this answers is not in the document]
Comment (z3d1k) [reply flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/KinesisShardSplitWithAssignmentStatus.java:47]: Added

File: flink-connector-aws/flink-connector-aws-kinesis-streams/src/test/java/org/apache/flink/connector/kinesis/source/enumerator/KinesisStreamsSourceEnumeratorTest.java (hunk 14 above)
Comment (z3d1k) [reply flink-connector-aws/flink-connector-aws-kinesis-streams/src/test/java/org/apache/flink/connector/kinesis/source/enumerator/KinesisStreamsSourceEnumeratorTest.java]: Removed

File: flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/tracker/SplitTracker.java [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+    public void markAsAssigned(Collection<KinesisShardSplit> splitsToAssign) {
+        splitsToAssign.forEach(split -> assignedSplits.add(split.splitId()));
+    }
+
+    /**
+     * Mark splits with specified ids as finished.
+     *
+     * @param finishedSplitIds collection of split ids to mark as finished
+     */
+    public void markAsFinished(Collection<String> finishedSplitIds) {
```
[the remark this answers is not in the document]
Comment (z3d1k) [reply flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/tracker/SplitTracker.java:91]: Added tests to verify that no exception is thrown in this case. This will be a no-op.

File: flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/tracker/SplitTracker.java [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+    }
+
+    /**
+     * Mark splits as assigned. Assigned splits will no longer be returned as pending splits.
+     *
+     * @param splitsToAssign collection of splits to mark as assigned
+     */
+    public void markAsAssigned(Collection<KinesisShardSplit> splitsToAssign) {
+        splitsToAssign.forEach(split -> assignedSplits.add(split.splitId()));
+    }
```
[the remark this answers is not in the document]
Comment (z3d1k) [reply flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/tracker/SplitTracker.java:84]: Added tests to verify that no exception is thrown in this case. This will be a no-op.

File: flink-connector-aws/flink-connector-aws-kinesis-streams/src/test/java/org/apache/flink/connector/kinesis/source/enumerator/tracker/SplitTrackerTest.java [hunk 20]
```diff
@@ -0,0 +1,387 @@
...
+        assertThatNoException().isThrownBy(() -> splitTracker.markAsFinished(splitIdsToFinish));
+    }
+
+    // -----------------------------------------------------------------------------------------
+    // Shard ordering disabled
+    // -----------------------------------------------------------------------------------------
+
+    @Test
+    @ValueSource(booleans = {true, false})
+    public void testUnorderedAllUnassignedSplitsAvailableForAssignment() {
```
Comment (hlteoh37) [inline flink-connector-aws/flink-connector-aws-kinesis-streams/src/test/java/org/apache/flink/connector/kinesis/source/enumerator/tracker/SplitTrackerTest.java:235]: the `@ValueSource` is not used

Comment (z3d1k) [reply flink-connector-aws/flink-connector-aws-kinesis-streams/src/test/java/org/apache/flink/connector/kinesis/source/enumerator/tracker/SplitTrackerTest.java:234]: removed

Comment (z3d1k) [reply flink-connector-aws/flink-connector-aws-kinesis-streams/src/test/java/org/apache/flink/connector/kinesis/source/enumerator/tracker/SplitTrackerTest.java:234]: removed

**Review by hlteoh37 — APPROVED** [review]
Thanks for the contribution @z3d1k :)
